### ISSUE TITLE
Running validation evaluation always failed when training the model

### ISSUE THREAD

**Issue by Shengqiang-Zhang** [rloganiv/kglm-model#48]
Thanks for the author's great work.
When I trained the kglm model, the validation evaluation always failed.
I use the following config:

> {
    "dataset_reader": {
        "type": "enhanced-wikitext-kglm",
        "alias_database_path": "data/linked-wikitext-2/alias.pkl"
    },
    "iterator": {
        "type": "fancy",
        "batch_size": 128,
        "split_size": 70,
        "splitting_keys": [
            "source",
            "target",
            "mention_type",
            "raw_entity_ids",
            "entity_ids",
            "parent_ids",
            "relations",
            "shortlist_inds",
            "alias_copy_inds"
        ]
    },
    "model": {
        "type": "kglm",
        "alias_encoder": {
            "type": "lstm",
            "hidden_size": 400,
            "input_size": 400
        },
        "cutoff": 30,
        "entity_embedder": {
            "token_embedders": {
                "entity_ids": {
                    "type": "embedding",
                    "embedding_dim": 256,
                    "pretrained_file": "data/linked-wikitext-2/embeddings.entities.txt",
                    "trainable": false,
                    "vocab_namespace": "entity_ids"
                }
            }
        },
        "hidden_size": 1150,
        "initializer": [
            [
                "token_embedder.weight",
                {
                    "a": -0.1,
                    "b": 0.1,
                    "type": "uniform"
                }
            ],
            [
                "decoder.bias",
                {
                    "type": "constant",
                    "val": 0
                }
            ]
        ],
        "knowledge_graph_path": "data/linked-wikitext-2/knowledge_graph.pkl",
        "num_layers": 3,
        "relation_embedder": {
            "token_embedders": {
                "relations": {
                    "type": "embedding",
                    "embedding_dim": 256,
                    "pretrained_file": "data/linked-wikitext-2/embeddings.relations.txt",
                    "trainable": true,
                    "vocab_namespace": "relations"
                }
            }
        },
        "tie_weights": true,
        "token_embedder": {
            "token_embedders": {
                "tokens": {
                    "type": "embedding",
                    "embedding_dim": 400,
                    "trainable": true
                }
            }
        },
        "use_shortlist": false
    },
    "train_data_path": "data/linked-wikitext-2/train.jsonl",
    "validation_data_path": "data/linked-wikitext-2/valid.jsonl",
    "test_data_path": "data/linked-wikitext-2/test.jsonl",
    "trainer": {
        "type": "lm",
        "cuda_device": [
            2,
            3
        ],
        "num_epochs": 1,
        "optimizer": {
            "type": "adam",
            "lr": 0.0003,
            "weight_decay": 1.2e-06
        },
        "patience": 10,
        "validation_metric": "-ppl"
    },
    "vocabulary": {
        "type": "extended",
        "directory_path": "data/linked-wikitext-2/vocab",
        "extend": false
    },
    "evaluate_on_test": true,
    "validation_dataset_reader": {
        "type": "enhanced-wikitext-kglm",
        "alias_database_path": "data/linked-wikitext-2/alias.pkl"
    },
    "validation_iterator": {
        "type": "fancy",
        "batch_size": 128,
        "split_size": 70,
        "splitting_keys": [
            "source",
            "target",
            "mention_type",
            "raw_entity_ids",
            "entity_ids",
            "parent_ids",
            "relations",
            "shortlist_inds",
            "alias_copy_inds"
        ]
    }
}

but the evaluation on validation set after each epoch always failed. The metric result as follows:

>  {
  "best_epoch": 0,
  "peak_cpu_memory_MB": 8197.112,
  "peak_gpu_0_memory_MB": 10,
  "peak_gpu_1_memory_MB": 10,
  "peak_gpu_2_memory_MB": 957,
  "peak_gpu_3_memory_MB": 10,
  "training_duration": "00:20:00",
  "training_start_epoch": 0,
  "training_epochs": 0,
  "epoch": 0,
  "training_ppl": 995.1405480875865,
  "training_upp": 1254.7674796458832,
  "training_kg_ppl": 193.04238067628987,
  "training_bg_ppl": 1197.0806547174739,
  "training_type": 0.46333958114065776,
  "training_new": 1.0831616436562888,
  "training_kg": 0.329283076343013,
  "training_vocab": 0,
  "training_new_p": 0.1187648456057007,
  "training_new_r": 0.0007927447995941146,
  "training_new_f1": 0.0015749767690913384,
  "training_kg_p": 0.0,
  "training_kg_r": 0.0,
  "training_kg_f1": 0.0,
  "training_new_ent_acc": 0.0,
  "training_new_ent_acc_20": 0.0,
  "training_parent_ppl": 4.633376004674751,
  "training_relation_ppl": 310.0364866986571,
  "training_loss": 8.898253054735138,
  "training_cpu_memory_MB": 8197.112,
  "training_gpu_0_memory_MB": 10,
  "training_gpu_1_memory_MB": 10,
  "training_gpu_2_memory_MB": 957,
  "training_gpu_3_memory_MB": 10,
  "validation_ppl": 1.0,
  "validation_upp": 1.0,
  "validation_kg_ppl": 1.0,
  "validation_bg_ppl": 1.0,
  "validation_type": 0,
  "validation_new": 0,
  "validation_kg": 0,
  "validation_vocab": 0,
  "validation_new_p": 0.0,
  "validation_new_r": 0.0,
  "validation_new_f1": 0.0,
  "validation_kg_p": 0.0,
  "validation_kg_r": 0.0,
  "validation_kg_f1": 0.0,
  "validation_new_ent_acc": 0.0,
  "validation_new_ent_acc_20": 0.0,
  "validation_parent_ppl": 1.0,
  "validation_relation_ppl": 1.0,
  "validation_loss": 0.0,
  "best_validation_ppl": 1.0,
  "best_validation_upp": 1.0,
  "best_validation_kg_ppl": 1.0,
  "best_validation_bg_ppl": 1.0,
  "best_validation_type": 0,
  "best_validation_new": 0,
  "best_validation_kg": 0,
  "best_validation_vocab": 0,
  "best_validation_new_p": 0.0,
  "best_validation_new_r": 0.0,
  "best_validation_new_f1": 0.0,
  "best_validation_kg_p": 0.0,
  "best_validation_kg_r": 0.0,
  "best_validation_kg_f1": 0.0,
  "best_validation_new_ent_acc": 0.0,
  "best_validation_new_ent_acc_20": 0.0,
  "best_validation_parent_ppl": 1.0,
  "best_validation_relation_ppl": 1.0,
  "best_validation_loss": 0.0
}

So I want to know how to config to get the evaluation on validation set work?

**Comment by rloganiv** [comment]
This issue is due to the fact that your validation iterator batch size (`128`) is larger than the number of documents in the validation data (`60`).

To explain in more detail, the `FancyIterator` is designed handle the issue where  
```
|-A-|-A-|-A-|-A-|-B-|...
|-C-|pad|pad|pad|-D-|...
```
In this case there will be many chunks which consist mainly of padding, and training will be biased to overfit the long sequence (e.g. sequence A above).

To avoid excess padding, it would be best to start a new sequence as soon as the previous one terminates, e.g.

|-A-|-A-|-A-|-A-|-B-|...
|-C-|-D-|...
    ^ D starts as soon as C ends

This is handled two ways:
1. Once a document is exhausted the `FancyIterator` will start yielding splits from a new document in the next batch.
